### DISCUSSION TITLE
v0.23.0-rc2: equivalent of CollectionId:  "_pb_users_auth_",

### DISCUSSION THREAD

**Discussion by Plorenzo** [pocketbase/pocketbase#5616] [Q&A] [answered]
I create my collections via migrations, I used to do this to create a relation to the users table. 

	`CollectionId:  "_pb_users_auth_"`

Now it looks like the users table id is autogenerated and I don't see any option to get the Auth collection Id from `core`

**Comment by ganigeorgiev** [comment]
You can still have custom collection id if you assign it manually, aka. `Collection.Id = ...`.

The new collection id when it is not explicitly set is NOT random (that actually was the old behavior) - now it is crc32 checksum of the name.

If you didn't set an explicit id and want to retrieve the generated id from another migration then you can send a query to retrieve the collection by its name, aka. `app.FindCollectionByNameOrId("test")`.

If that doesn't help, please elaborate a little more by providing a more complete code sample.

**Comment by Plorenzo** [comment]
Sorry I didn't explain myself very well. 

In the old version the users table had consistently this id `_pb_users_auth_`, so you could use it to create a new collection and create a relation filed to the auth table using that id, like so

```go
func init() {
	m.Register(func(db dbx.Builder) error {
		collection := &models.Collection{
			Name:       "favourites",
			Type:       models.CollectionTypeBase,
			ListRule:   types.Pointer("user = @request.auth.id"),
			ViewRule:   types.Pointer("user = @request.auth.id"),
			CreateRule: types.Pointer("user = @request.auth.id && @request.auth.id != listing.user"),
			UpdateRule: types.Pointer("user = @request.auth.id"),
			DeleteRule: types.Pointer("user = @request.auth.id"),
			Schema: schema.NewSchema(
				&schema.SchemaField{
					Name:     "user",
					Type:     schema.FieldTypeRelation,
					Required: true,
					Unique:   true,
					Options: &schema.RelationOptions{
						CollectionId:  "_pb_users_auth_",
						CascadeDelete: true,
						MaxSelect:     types.Pointer(1),
					},
				},
			),
		}
		collection.SetId("favourites_id")
		if err := daos.New(db).SaveCollection(collection); err != nil {
			return err
		}
		return nil
	}, func(db dbx.Builder) error {
		// add down queries...
		return nil
	})
}
```

now, as you mention the id has changed and it's rc32 checksum of the name, so I was asking how to retrieve this value for the users table. 

I guess now I need to use `app.FindCollectionByNameOrId("users")` like this 
```go

                coll, _ := app.FindCollectionByNameOrId("users")

		collection.Fields.Add(&core.RelationField{
			Name:          "user",
			Required:      true,
			MaxSelect:     1,
			CollectionId:  coll.Id,
			CascadeDelete: true,
		})
```

or will the rc32 value consistently  be `_pbc_344172009` and we can skip 1 query?

**Reply by ganigeorgiev** [reply, marked as answer]
> will the rc32 value consistently be _pbc_344172009 and we can skip 1 query?

The collection id will be crc32 checksum of the name only if the `Id` field is not explicitly set.

If you are concerned about performance - don't, there is no network request and the migrations are executed only once during startup.

A translated version of the above migration could look something like:

```go
package migrations

import (
	"github.com/pocketbase/pocketbase/core"
	m "github.com/pocketbase/pocketbase/migrations"
	"github.com/pocketbase/pocketbase/tools/types"
)

func init() {
	m.Register(func(app core.App) error {
		usersCollection, err := app.FindCollectionByNameOrId("users")
		if err != nil {
			return err
		}

		collection := core.NewBaseCollection(("favourites"))
		collection.Id = "favourites_id"
		collection.ListRule = types.Pointer("user = @request.auth.id")
		collection.ViewRule = types.Pointer("user = @request.auth.id")
		collection.CreateRule = types.Pointer("user = @request.auth.id && @request.auth.id != listing.user")
		collection.UpdateRule = types.Pointer("user = @request.auth.id")
		collection.DeleteRule = types.Pointer("user = @request.auth.id")
		collection.Fields.Add(&core.RelationField{
			Name:          "user",
			Required:      true,
			CollectionId:  usersCollection.Id,
			CascadeDelete: true,
			MaxSelect:     1, // can be skipped because with v0.23.0 "0" or "1" value means single relation
			// note that the Unique field option is no-op since PocketBase v0.14 and instead
			// separate UNIQUE index for the field is required
			// (https://github.com/pocketbase/pocketbase/blob/master/CHANGELOG_8_15.md#v0140)
		})
		// add the index using plain string appended to collection.Indexes or use the model helpers
		collection.AddIndex("idx_favourites_user", true, "user", "")

		return app.Save(collection)
	}, nil)
}
```

**Reply by ganigeorgiev** [reply]
_As a side-note, if all of your migrations are schema related then you may find it easier to migrate by deleting all of your migration files, updgrade your other code and then run `go run . migrate collections` to generate a complete initial snapshot of your auto upgraded collections configuration (expected to be run on the old pb_data so that it can properly include the old&migrated state). The old `pb_data` and its collections configuration is automigrated by the system [`1717233556_v0.23_migrate.go`](https://github.com/pocketbase/pocketbase/blob/develop/migrations/1717233556_v0.23_migrate.go) (the timestamp may change with the final v0.23.0 release)_

**Reply by Plorenzo** [reply]
thanks! not concerned about performance as you mentioned it's just a 1 time op, just curious. Also, I write a lot of migrations code, so I rather do the migration manually to learn all the new stuff :) 

other thing I'm noticing is that migration order matter now but not in v0.22.21. Im getting this error, because a migration is trying to create a collection with a relation field that points to a collection that will be created in a subsequent migration. 

`Error: Failed to apply migration 1672225404_create_listings_collection.go: fields: (11: (collectionId: The relation collection doesn't exist..); 12: (collectionId: The relation collection doesn't exist..).).` 

In the old version not sure why, this was working fine,  even out of order migrations were getting executed correctly

**Reply by Plorenzo** [reply]
with in order I mean alphabetical order of the migration files 

```
1672323199_create_favourites_collection.go 
1672325404_create_listings_collection.go 
1673015496_create_messages_collection.go 
1673015513_create_conversations_collection.go 
1687092734_create_settings.go 
1689864303_create_admin.go 
1690020361_update_users.go 
1690040390_create_reviews.go
1702392151_create_reported_listings.go 
...
```

**Reply by ganigeorgiev** [reply]
> other thing I'm noticing is that migration order matter now but not in v0.22.21.

No, you are mistaken. Migrations order has always mattered. That's one of reason why we have the timestamp prefix.

The migrations can execute in a different sort order in case they were created and applied in different times. For example:

```
1. created migration "5_x.go"
2. "5_x.go" is applied automatically on app start because it is missing
3. someone else adds (merge) a new migration "4_x.go"
4. "4_x.go" is applied automatically on app start because it is missing
```

But if you init a new app with blank `pb_data`, then the migrations will be applied in the expected order, aka. "4_x.go", "5_x.go".

**Reply by Plorenzo** [reply]
mmmm maybe I'm missing something then, but in my project If I `rm -rf pb_data/` and run the app, the migrations get applied just fine despite being not in order. 

This is a simplified version of my project that runs just fine in v0.22.21 with migrations out of order. For instance `create_favourites ` creates a field pointing to `listings`, before `listings` is created. Same for `create_listing` and `boats` and `cancellationPolicies`
If you `go run main.go serve ` you will see it runs fine and the tables get created. 

[pbRepro.zip](https://github.com/user-attachments/files/17279892/pbRepro.zip)

this is the equivalent in v0.23.0-rc2
[pbRepro v0.23.zip](https://github.com/user-attachments/files/17279935/pbRepro.v0.23.zip)

and we run, I get: 

`Error: Failed to apply migration 1672323199_create_favourites_collection.go: createRule: Invalid rule. Raw error: invalid right operand "listing.user" - failed to load field "listing" collection; fields: (2: (collectionId: The relation collection doesn't exist..).).
`

It's totally possible that I'm missing something or doing something wrong, so forgive me if that's the case

**Reply by ganigeorgiev** [reply]
I don't have a sandboxed VM at the moment to download and explore the files.

Again, there is no difference in the migrations order execution from earlier versions.

What I think is causing the problem in your case is that in the old version `Dao.SaveCollection(collection)` doesn't run validations and this is no longer the case with `app.Save(collection)`.

If you want the old behavior and to insert the collection without validations, then you can run `app.SaveNoValidate(collection)`.
